File: snippets/depth-first-search/dfs.cpp
#include <bits/stdc++.h>

using namespace std;

class Graph {
public:
    int V;
    vector<list<int>> adj;

    Graph(int V) {
        this->V = V;
        adj.resize(V);
    }

    void AddEdge(int src, int dest) {
        adj[src].push_back(dest);
    }

    void AddBiEdge(int u, int v) {
        AddEdge(u, v);
        AddEdge(v, u);
    }

    void Dfs(int src) {
        vector<bool> visited(V, false);
        stack<int> dfs;

        dfs.push(src);

        while (!dfs.empty()) {
            int v = dfs.top();
            dfs.pop();

            if (!visited[v]) {
                visited[v] = true;
                cout << v << ' ';  ///
            }

            for (const auto& u : adj[v])
                if (!visited[u])
                    dfs.push(u);
        }

        cout << endl;  ///
    }

    void DfsRecur(int src) {
        vector<bool> visited(V, false);
        DfsUtil(src, visited);

        cout << endl;  ///
    }

    void DfsUtil(int v, vector<bool>& visited) {
        visited[v] = true;

        cout << v << ' ';  ///

        for (const auto& u : adj[v])
            if (!visited[u])
                DfsUtil(u, visited);
    }
};

File: snippets/depth-first-search/dfs.py
class Graph(object):
    def __init__(self, V):
        self.V = V
        self.adj = [[] for i in range(V)]

    def add_edge(self, src, dest):
        self.adj[src].append(dest)

    def add_bi_edge(self, u, v):
        self.add_edge(u, v)
        self.add_edge(v, u)

    def dfs(self, src):
        visited = [False for i in range(self.V)]
        dfs = []

        dfs.append(src)

        while len(dfs) > 0:
            v = dfs.pop()

            if not visited[v]:
                visited[v] = True
                print(v, end=' ')  ###

            for u in self.adj[v]:
                if not visited[u]:
                    dfs.append(u)

        print()

    def dfs_recur(self, src):
        visited = [False for i in range(self.V)]
        self.dfs_util(src, visited)
        print()  ###

    def dfs_util(self, v, visited):
        visited[v] = True

        print(v, end=' ')  ###

        for u in self.adj[v]:
            if not visited[u]:
                self.dfs_util(u, visited)
